### PR TITLE
feat: added mime accept headers

### PR DESCRIPTION
**Pull request by StephenHodgson**
This ensures that returned payload is in the format that a client can accept.

- Added way to specify accept header mime types for returned image and audio content in the specified format
- Added unit tests for text-to-speech, audio-to-audio, and text-to-image tasks

### CODE REVIEW

**Comment by Narsil** [comment]
Which formats do you actually need ? I'd rather whitelist a few mimetypes than accepting arbitrary ones.

**Comment by StephenHodgson** [comment]
We should at minimum use the ones outlined in the [inference serialization docs](https://huggingface.co/docs/inference-endpoints/others/serialization)

<html>
<body>


Accept | Payload
-- | --
application/json | JSON
text/csv | raw
text/plain | raw
image/png | binary
image/jpeg | binary
image/jpg | binary
image/tiff | binary
image/bmp | binary
image/gif | binary
image/webp | binary
image/x-image | binary
audio/x-flac | binary
audio/flac | binary
audio/mpeg | binary
audio/x-mpeg-3 | binary
audio/wave | binary
audio/wav | binary
audio/x-wav | binary
audio/ogg | binary
audio/x-audio | binary
audio/webm | binary
audio/webm;codecs=opus | binary
audio/AMR | binary
audio/amr | binary
audio/AMR-WB | binary
audio/AMR-WB+ | binary
audio/m4a | binary
audio/x-m4a | binary


</body>
</html>

**Comment by StephenHodgson** [comment]
> Which formats do you actually need ? I'd rather whitelist a few mimetypes than accepting arbitrary ones.

Besides the fact that it doesn't _actually_ affect anything except for audio-to-audio endpoints and text-to-speech endpoints, I think this is relatively low risk change.

**Comment by Narsil** [comment]
As you can see most proposed files here are not even audio, which is why whitelisting is better imo.
It also prevents injection (even though we should be fine in the case, in general passing things as-is is not great).

But let me rephrase the question, what is the actual problem you are encountering ? I don't like doing modifications without knowing the root cause of the issue.

On the browser flac seems supported by all major browser vendors directly, so the hub doesn't require any re-encoding afaik.
Flac is lossless and smaller than wav, so quality perception cannot be linked to the encoding of the audio (which is good since we don't know in what domain the actual audio is going to be, so by choosing other encoding we might add artifacts not generated by the model).

**Comment by StephenHodgson** [comment]
> But let me rephrase the question, what is the actual problem you are encountering ? I don't like doing modifications without knowing the root cause of the issue.

I've already outlined it in the body of the PR description:
> This ensures that returned audio is in formats that is accepted by api clients that cannot convert or re-encode locally.

**Comment by StephenHodgson** [comment]
> in general passing things as-is is not great

I mean literally all the inputs and parameters are also unfiltered and passed as is 😏

**Comment by StephenHodgson** [comment]
@Narsil I added a mime validation function

File: api_inference_community/routes.py [hunk 1]
```diff
@@ -27,6 +28,10 @@ async def pipeline_route(request: Request) -> Response:
     start = time.time()
     payload = await request.body()
     task = os.environ["TASK"]
+    accept = request.headers["accept"]
+    # Parse accept header and determine the appropriate format
```
Comment (StephenHodgson) [inline api_inference_community/routes.py:32]:
```suggestion
    # Parse accept header and determine the appropriate response format
```

**Comment by StephenHodgson** [comment]
@Narsil I did some refactoring. Let me know what you think

File: api_inference_community/routes.py [hunk 2]
```diff
@@ -35,6 +37,7 @@ async def pipeline_route(request: Request) -> Response:
             sampling_rate = pipe.sampling_rate
         except Exception:
             sampling_rate = None
+        mime_types= validate_mime_types(request.headers["accept"])
```
Comment (StephenHodgson) [inline api_inference_community/routes.py:40]:
```suggestion
        mime_types = validate_mime_types(request.headers["accept"])
```

**Comment by StephenHodgson** [comment]
@Narsil please have a look at the updated implementation

**Comment by Narsil** [comment]
> What makes you assume I'm targeting a browser for my api client?

This is what this repo is made for. Enabling the widgets for as many models as possible *for free*. It is not made for professional offerings.
For professional offering with full control we recommend using spaces or https://huggingface.co/inference-endpoints

> This ensures that returned audio is in formats that is accepted by api clients that cannot convert or re-encode locally.

Sorry, this is not a problem. We're not providing the API as a free converter service.
A problem would be some browser/OS/vendor failing to read the proposed file, which would indeed warrant a good ground to provide a fix. Knowing the actual issue could help narrow down the actually necessary file formats.

> The docs state that the API supports setting the Accept mime content types, but it is not actually implemented/supported

Where exactly ?

**Comment by NonpareilNic** [comment]
@StephenHodgson sounds like Fork territory to me!

**Comment by StephenHodgson** [comment]
> Sorry, this is not a problem. We're not providing the API as a free converter service.
> A problem would be some browser/OS/vendor failing to read the proposed file, which would indeed warrant a good ground to provide a fix. Knowing the actual issue could help narrow down the actually necessary file formats.

I'm simply trying to get the returned data in a format that my rest client can easily understand. I'm not trying to run up the costs on your "basic free services". [Besides the fact, it already seems like you're doing a conversion in the first place to the flac format](https://github.com/huggingface/api-inference-community/blob/c08dd1d13baae01440299d61f46fff67ddcfaa26/api_inference_community/validation.py#LL267C12-L267C12).
Which is why I don't understand why you are making this such a big deal.

Funny enough the examples provided by [the inference client docs](https://github.com/huggingface/huggingface_hub/blob/8eabd33e4620d6246eb7f1ba15646f784384e0f6/src/huggingface_hub/_inference.py#L763) seem to suggest that you can write a file directly to `.wav`, even though the returned data is actually encoded as `flac`.

```py
>>> from pathlib import Path
>>> from huggingface_hub import InferenceClient
>>> client = InferenceClient()
>>> audio = client.text_to_speech("Hello world")
>>> Path("hello_world.wav").write_bytes(audio)
```
I tried this out locally with the python InferenceClient and indeed VLC told me the encoded file was actually `flac`!
image

I am kinda curious as to the original format it is converting _from_. If it was raw PCM data, it would be great because then no conversion would need to happen at all.

I simply want to be able to set the accept header to the specified format and the returned content-type be something that my rest client can understand without having to do an additional conversion process on the client side. This seems like a very straight forward ask, [since most RESTful client/server relationships typically provide this sort of mechanism](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept).

> The Accept request HTTP header indicates which content types, expressed as [MIME types](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types), the client is able to understand.

> This is what this repo is made for. Enabling the widgets for as many models as possible for free. It is not made for professional offerings. For professional offering with full control we recommend using spaces or https://huggingface.co/inference-endpoints

Sure, but doesn't the python [InferenceClient](https://huggingface.co/docs/huggingface_hub/v0.15.1/en/package_reference/inference_client#huggingface_hub.InferenceClient) actually call these endpoints under the hood? That means other Hugging Face projects [like this one](https://github.com/huggingface/unity-api) cannot fully utilize all the endpoints provided by this resource. I'm trying to get this completed before the [Hugging Face Game Jam](https://itch.io/jam/open-source-ai-game-jam) so that all of the tasks supported by the widgets are available for jammers to use.

**Comment by Narsil** [comment]
> > Sorry, this is not a problem. We're not providing the API as a free converter service.
> 
> First off, wow, very rude. I've taken time to contribute back for everyone's benefit, not just my own.

You're entirely correct. Please accept my apologies for being rude.
I tend to be direct in general in written form, but this was not warranted for.

I do think you mean well, but I think here there might be misalignment between what this repo is trying to achieve and what you're trying to do.

> I'm simply trying to get the returned data in a format that my rest client can easily understand. I'm not trying to run up the costs on your "basic free services". [Besides the fact, it already seems like you're doing a conversion in the first place to the flac format](https://github.com/huggingface/api-inference-community/blob/c08dd1d13baae01440299d61f46fff67ddcfaa26/api_inference_community/validation.py#LL267C12-L267C12). Which is why I don't understand why you are making this such a big deal.

The cost is mainly a maintenance developer cost. If we support that, for every issue people are going to see we now need to think what was the format that the thing actually sent, which we have no way of knowing.

> I am kinda curious as to the original format it is converting _from_. If it was raw PCM data, it would be great because then no conversion would need to happen at all.

For ML, everything is wav (PCM f32) in the end (usually created from mel spectro).
Raw f32 PCM is also much larger and incurs larger network latency, which is not ideal (hence the choice of flac, mp3 was discarded because of potential artifacts that could be introduced)

> 
> Funny enough the examples provided by [the inference client docs](https://github.com/huggingface/huggingface_hub/blob/8eabd33e4620d6246eb7f1ba15646f784384e0f6/src/huggingface_hub/_inference.py#L763) seem to suggest that you can write a file directly to `.wav`, even though the returned data is actually encoded as `flac`.

Indeed, sorry I do not maintain this part, and indeed the `.wav` extension is misleading. @Wauplin Do you think we can modify this ?

In general the API itself should send back the proper mimetype (I don't think it's exposed by the client).

> I simply want to be able to set the accept header to the specified format and the returned content-type be something that my rest client can understand without having to do an additional conversion process on the client side.

I know, and I know you mean well. But accepting and maintaining this is also a pain. And since everything provided here is for demo purposes (every advanced usage of models is handled elsewhere in our stack), it's really beneficial to keep things simple here. Maintaining 300k models over 20 frameworks over 20 tasks is complex enough.
Currently flac is a single format which seems to solve every issue we have and try to solve.

If you re-expose audio to some other things that do not talk flac, sorry but I'd like to know which ones, and whitelist them one by one. raw PCM for instance is never Raw, there's a lot of added data into sampling size and order, so it would most likely be WAV, and WAV creates larger files, which occur network costs (as in latency cost). MP3 is smaller and degrades quality.

But every surface added here is surface that has to be maintained, and that could potentially create unintended consequences. Like degrading quality of a given model because of incorrect encoding.

> Sure, but doesn't the python [InferenceClient](https://huggingface.co/docs/huggingface_hub/v0.15.1/en/package_reference/inference_client#huggingface_hub.InferenceClient) actually call these endpoints under the hood? That means other Hugging Face projects [like this one](https://github.com/huggingface/unity-api) cannot fully utilize all the endpoints provided by this resource.
I think `InferenceClient` should cover all our endpoints. @Wauplin should be able to confirm.


>  I'm trying to get this completed before the [Hugging Face Game Jam](https://itch.io/jam/open-source-ai-game-jam) so that all of the tasks supported by the widgets are available for jammers to use.

Thanks for providing context this definitely helps a lot getting a perspective of what you're trying to achieve.
So my guess is that `flac` is not as standard as you'd like in games, and you would like to see something more suitable, correct ? What would be the necessary ? Would `wav` be enough.

As mentionned earlier, I'm very sensitive to adding huge lists of mimetypes and sending things as-is to some other binary. Both from a security and a maintenance perspective it's really easy to create nasty situations.

> Fork territory to me!

By all means please do, this is what open source is great for. This repo is made for wide support and very simple functionality. Definitely not optimal by any means, and features are kept to a minimum on purpose (hence the pushback on adding support for a feature without a properly described issue).

**Comment by Wauplin** [comment]
> Indeed, sorry I do not maintain this part, and indeed the .wav extension is misleading. @Wauplin Do you think we can modify this ?

Yes definitely. I assumed it was a .wav but didn't look into the details when implementing it. I'll make a PR to fix the doc.

> I think InferenceClient should cover all our endpoints.

Yes `InferenceClient` is meant for Inference API (this repo), [Inference Endpoints](https://huggingface.co/inference-endpoints) or any API endpoint that runs with the same specs. But I did not spent time on handling custom headers including mime-type yet -but that's a bit orthogonal to the discussion here-.

**Comment by StephenHodgson** [comment]
# Short Reply

> Thanks for providing context this definitely helps a lot getting a perspective of what you're trying to achieve.
So my guess is that flac is not as standard as you'd like in games, and you would like to see something more suitable, correct ? What would be the necessary ? Would wav be enough.

Yes `.wav` would be something we could use, `.ogg` even better since it's small and easier to load into memory.

@Narsil please give me clear guidance about what changes you want me to make to the PR to satisify any constraints so that I can get this done and merged 😃

> Fork territory to me!

Lol, no I want to contribute here so that the Hugging Face Game Jammers can take advantage of what is already offered.

---

# Continuing the longer discussion about mime types:

> Maintaining 300k models over 20 frameworks over 20 tasks is complex enough.

👏 No small feat! 🎉

>  But I did not spent time on handling custom headers including mime-type yet -but that's a bit orthogonal to the discussion here-.

I do want to point out that the PR doesn't really modify anything other than the audio conversions, and that the other MIME types I put in the enum are what is specified as valid mime types by the inference API docs [here](https://huggingface.co/docs/inference-endpoints/others/serialization).

I think having at minimum the ability to handle the mime-types for at least the outlined in the docs would be nice.

> since everything provided here is for demo purposes (every advanced usage of models is handled elsewhere in our stack)

It is worth mentioning that even with private models/repos, this repo does in fact as as the default API mechanism unless a [custom handler is implemented](https://huggingface.co/docs/inference-endpoints/guides/custom_handler).

So either the responsibilities of this repo has creeped in scope, or the paradigm needs to be readdressed and a proper solution put in place. If it meant to be a demo, that's totally fine! It is pretty much what I assumed with wording around the free plan and rate limiting outlined in the docs.

I do intend to take my client a bit further in the future and add implementations for the higher cost Inference Endpoints, but since those actually support customizing the handler, it'll be much more nuanced per endpoint. (Although I'm sure there should be a way to enforce the custom handler to return a defined schema, as an extension of the OpenAPI.json)

Maybe a more in depth discussion in a chat channel or forum about this topic may be beneficial to get feedback from the community at large.

File: api_inference_community/routes.py [hunk 3]
```diff
@@ -35,6 +37,7 @@ async def pipeline_route(request: Request) -> Response:
             sampling_rate = pipe.sampling_rate
         except Exception:
             sampling_rate = None
+        mime_types = validate_mime_types(request.headers["accept"])
```
Comment (Narsil) [inline api_inference_community/routes.py:40]: Ok, my suggestion:

We can add a new `parameter` which would be `audio_format` maybe ? (There might be a  better name).

We can whitelist it to be `flac`, `mp3`, `ogg`, `wav` so that we cover most used formats I think.

That means users can choose encoding directly from params. We can then respect `headers["accept"]` to override the parameter (if it hasn't been specified).

Because it's a parameter, there's much less code change in other places.

File: api_inference_community/routes.py [hunk 4]
```diff
@@ -94,26 +97,36 @@ def emit(self, record):
...
             waveform, sampling_rate = outputs
-            data = ffmpeg_convert(waveform, sampling_rate)
-            headers["content-type"] = "audio/flac"
+
+            # Use the first valid mime type and its corresponding extension for the conversion format or "flac" as default
+            content_type, ext = list(mime_types.items())[0] if mime_types else ("audio/flac", "flac")
+            format_for_conversion = ext.lstrip(".")
+
+            data = ffmpeg_convert(waveform, sampling_rate, format_for_conversion)
+            headers["content-type"] = content_type
```
Comment (Narsil) [inline api_inference_community/routes.py:112]: This could simply become something like:

```python
format_for_conversion = params.get("audio_format", "flac")
data = ffmpeg_convert(waveform, sampling_rate, format_for_conversion)
headers["content-type"] = CONTENT_TYPE[format_for_conversion]
```
`CONTENT_TYPE` being a static map of ffmpeg names to content-types.

File: api_inference_community/validation.py [hunk 5]
```diff
@@ -172,6 +176,13 @@ class StringInput(BaseModel):
     "zero-shot-classification": ZeroShotParamsCheck,
 }
 
+
+def check_params(params, tag):
+    if tag in PARAMS_MAPPING:
+        PARAMS_MAPPING[tag].parse_obj(params)
+    return True
+
```
Comment (Narsil) [inline api_inference_community/validation.py:184]: Can you revert those changes, they don't seem linked to the PR at hand.

File: api_inference_community/validation.py [hunk 6]
```diff
@@ -209,13 +212,54 @@ def check_inputs(inputs, tag):
...
+    "audio/ogg",
+    "audio/x-audio",
+    "audio/webm;codecs=opus"
+    "audio/AMR",
+    "audio/amr",
+    "audio/AMR-WB",
+    "audio/AMR-WB+",
+    "audio/m4a",
+    "audio/x-m4a"
+}
```
Comment (Narsil) [inline api_inference_community/validation.py:243]: Let's remove all those and keep only the suggested ones

File: api_inference_community/validation.py [hunk 7]
```diff
@@ -209,13 +212,54 @@ def check_inputs(inputs, tag):
...
+
+
+def validate_mime_types(accept_header: Any) -> Dict[str, str]:
+    mime = MimeTypes()
+    mime_dict = {mime_type: mime.guess_extension(mime_type) for mime_type in accept_header.split(',')}
+    for mime_type in mime_dict:
+        if mime_type not in MIME_TYPES:
+            raise ValueError(f"{mime_type} is not a supported MIME type.")
```
Comment (Narsil) [inline api_inference_community/validation.py:251]: I think ignoring here would be better than raising. We should raise only if ALL accept mimetypes are invalid, no ?

File: api_inference_community/validation.py [hunk 8]
```diff
@@ -209,13 +212,54 @@ def check_inputs(inputs, tag):
...
+    return mime_dict
```
Comment (Narsil) [inline api_inference_community/validation.py:252]: I think it would be nice if we returned only the actually resolved `format_for_conversion`.

File: api_inference_community/validation.py [hunk 9]
```diff
@@ -209,13 +212,54 @@ def check_inputs(inputs, tag):
...
+
+
```
Comment (Narsil) [inline api_inference_community/validation.py:254]: We need to update `AUDIO_INPUTS` to add the format `audio_format` parameter (and validation).

**Review by Narsil — COMMENTED** [review]
I left a few comments about the approach I suggest.

What I really care for:
- Making the PR as small as possible (not to an extreme, but types changes in Python tend to cause issues, so modifying signatures of existing functions is always a pain)
- Whitelisting types to a handful. (I don't have huge preferences, but PNG seems wrong for audio).
- Adding a few tests. Mostly about making sure that sending invalid headers and/or invalid params actually raises errors. Existing test should be easy to copy, if not I am happy to jump in to check. 

Ideally we would check the actual end conversion but currently this crate doesn't check ffmpeg behavior (iirc)

Note: Changing this library, will not automatically change the actual inference code of various libraries. We will need to update `api-inference-community` version on the intended frameworks for it to work. We need to make 1 PR per framework afterwards. (I  need to manually create a new version of `api-inference-community`, there is no automated builds currently, but it's super fast)

File: api_inference_community/routes.py (hunk 3 above)
Comment (StephenHodgson) [reply api_inference_community/routes.py]: We could even make it more generic and just have optional `format` for any pipeline

**Comment by StephenHodgson** [comment]
@Narsil @Wauplin please have a look at the current implementation. 

I've updated the `normalize_payload` function to assign an `audio_format` and defaults to `flac`

This can also be modified in the future to change image encodings of accepted types in the future as well if the need ever arises.

**Comment by StephenHodgson** [comment]
We already have this list of `AUDIO_EXTENSIONS` I considered using that instead of adding new whitelisted dict.

File: api_inference_community/validation.py [hunk 10]
```diff
@@ -244,18 +249,56 @@ def check_inputs(inputs, tag):
...
+    "audio/wav": "wav",
+    "audio/ogg": "ogg",
+    "audio/mp4": "m4a",
+    "audio/aac": "aac",
+    "audio/webm": "webm",
+    "image/jpeg": "jpg",
+    "image/png": "png",
+    "image/bmp": "bmp",
+    "image/tiff": "tiff",
+    "image/webp": "webp",
```
Comment (Narsil) [inline api_inference_community/validation.py:264]: Can we split this map into 2 ? 1 for audio 1 for images ? Currently a user could ask tiff on mimetypes making the thing crash on ffmpeg.

Splitting into 2 could also help remove the `image/` /  `audio/` prefix (which you are readding later anyway, so we could drop from dict to set leading to easier lookups (instead of doing seq scans in dict.values()).

The perf doesn't matter here, but the code should end up being cleaner

File: api_inference_community/validation.py [hunk 11]
```diff
@@ -244,18 +249,56 @@ def check_inputs(inputs, tag):
...
+}
+
+
 def normalize_payload(
-    bpayload: bytes, task: str, sampling_rate: Optional[int] = None
+    bpayload: bytes, task: str, sampling_rate: Optional[int] = None, accept_header: Optional[Any] = None
```
Comment (Narsil) [inline api_inference_community/validation.py:269]:
```suggestion
    bpayload: bytes, task: str, sampling_rate: Optional[int] = None, accept_header: Optional[str] = None
```
No ?

**Review by Narsil — APPROVED** [review]
LGTM.

I left a nit for splitting the whistlist between audio and images. I can take care of it for you if you want.

File: api_inference_community/validation.py (hunk 10 above)
Comment (StephenHodgson) [reply api_inference_community/validation.py:264]: >  Currently a user could ask tiff on mimetypes making the thing crash on ffmpeg.

oof, good catch, I also wanna write a unit test for this.

Comment (StephenHodgson) [reply api_inference_community/validation.py]: @Narsil I split up the map as suggested

**Comment by StephenHodgson** [comment]
@Narsil how can I run tests locally?

**Comment by StephenHodgson** [comment]
I might need some help with the tests

**Comment by StephenHodgson** [comment]
Just tested this out locally. Is it live already?

The files are still coming back encoded as flac

**Comment by Narsil** [comment]
No, every library has their own version of `api-inference-community`. So we need to update the libs. And then, models are not automatically cycled, so if some model is highly used, it's never turned off, and never turned back on with an earlier version.

Which models do you need ?

**Comment by StephenHodgson** [comment]
I suppose just the recommended models for text to speech and audio to audio tasks:

- Text to Speech: `espnet/kan-bayashi_ljspeech_vits`
- Audio to Audio: `speechbrain/sepformer-wham`

In the meantime I'll just add a warning to the client side to notify them that the model api may not be up to date.

**Comment by Narsil** [comment]
No, there is not.
